algorithm/cgal/cgal.cc

```diff
@@ -4779,6 +4695,105 @@ void FromSurfaceMeshToPolygonsWithHoles(const Surface_mesh* input,
+                                           const Transformation* transform,
+                                           double offset, int segments,
+                                           double threshold,
+                                           emscripten::val emit_polygon,
+                                           emscripten::val emit_point) {
+  namespace PS = CGAL::Polyline_simplification_2;
+  typedef CGAL::Polygon_2<Kernel> Polygon_2;
+  typedef CGAL::Polygon_with_holes_2<Kernel> Polygon_with_holes_2;
+  typedef PS::Vertex_base_2<Kernel> Vb;
+  typedef CGAL::Constrained_triangulation_face_base_2<Kernel> Fb;
+  typedef CGAL::Triangulation_data_structure_2<Vb, Fb> TDS;
+  typedef CGAL::Constrained_Delaunay_triangulation_2<Kernel, TDS,
+                                                     CGAL::Exact_predicates_tag>
+      CDT;
+  typedef CGAL::Constrained_triangulation_plus_2<CDT> CT;
+  typedef CT::Vertices_in_constraint_iterator Vertices_in_constraint_iterator;
+  typedef PS::Stop_above_cost_threshold Stop;
+  typedef PS::Squared_distance_cost Cost;
+
+  Surface_mesh mesh(*input);
+  CGAL::Polygon_mesh_processing::transform(*transform, mesh,
+                                           CGAL::parameters::all_default());
+
+  std::vector<CT::Constraint_id> cids;
+  Plane xy(0, 0, 1, 0);
+
+  std::unordered_map<Plane, Arrangement_2> arrangements;
+  convertSurfaceMeshFacesToArrangements(mesh, arrangements,
+                                        /*use_unit_planes=*/true);
+  for (const auto& entry : arrangements) {
+    const Arrangement_2& arrangement = entry.second;
+    std::vector<Polygon_with_holes_2> polygons;
+    convertArrangementToPolygonsWithHoles(arrangement, polygons);
+
+    CT ct;
+    CT::Constraint_id null_constraint;
+    std::vector<Polygon_with_holes_2> offset_polygons;
+    for (const Polygon_with_holes_2& polygon : polygons) {
+      // These will form the unsimplified payload.
+      ct.insert_constraint(polygon.outer_boundary());
+      std::vector<Polygon_2> holes;
+      Polygon_2 boundary = polygon.outer_boundary();
+      offsetOfPolygonWithHoles(offset, -1, -1, segments, xy, boundary, holes,
+                               offset_polygons);
+    }
+    for (const Polygon_with_holes_2& polygon : offset_polygons) {
+      // These will form the simplified envelopes.
+      CT::Constraint_id cid = ct.insert_constraint(polygon.outer_boundary());
+      if (cid != null_constraint) {
+        cids.push_back(cid);
+      }
+    }
+
+    Cost cost;
+    Stop stop(threshold);
+
+    // Simplify each envelope.
+    for (CT::Constraint_id cid : cids) {
+      PS::simplify(ct, cid, cost, stop, true);
+    }
+
+    // And emit.
+    for (CT::Constraint_id cid : cids) {
+      emit_polygon(false);
+      Vertices_in_constraint_iterator it = ct.vertices_in_constraint_begin(cid);
+      Vertices_in_constraint_iterator end = ct.vertices_in_constraint_end(cid);
+      if (it != end) {
+        for (;;) {
```

Comment:
I had to look up how this works, it's neat. I didn't realize that ALL of the arguments to for() are optional

Reply:
It's not that they're optional, but rather that an empty statement is valid. :)